File: src/main/java/com/kmatrokhin/uvbot/services/UserService.java
package com.kmatrokhin.uvbot.services;

import com.kmatrokhin.uvbot.dto.Coordinates;
import com.kmatrokhin.uvbot.dto.LocationInfo;
import com.kmatrokhin.uvbot.dto.UserSignUp;
import com.kmatrokhin.uvbot.entities.LocationEntity;
import com.kmatrokhin.uvbot.entities.UserEntity;
import com.kmatrokhin.uvbot.entities.UserLanguage;
import com.kmatrokhin.uvbot.repositories.LocationRepository;
import com.kmatrokhin.uvbot.repositories.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.Instant;
import java.util.Optional;

@Service
@RequiredArgsConstructor
public class UserService {
    private final UserRepository userRepository;
    private final LocationRepository locationRepository;

    @Transactional
    public UserEntity signUpOrUpdate(UserSignUp userSignUp) {
        Long chatId = userSignUp.getChatId();
        String username = userSignUp.getName();
        LocationInfo locationInfo = userSignUp.getLocationInfo();

        Optional<UserEntity> userEntityOpt = userRepository.findByChatId(chatId);
        Coordinates coordinates = locationInfo.getCoordinates();
        UserEntity userEntity;
        if (userEntityOpt.isPresent()) {
            userEntity = userEntityOpt.get();
            userEntity
                .setName(username)
                .setIsSubscribed(true);
            locationRepository.getByUserEntity(userEntity)
                .setName(locationInfo.getName())
                .setLatitude(coordinates.getLatitude())
                .setLongitude(coordinates.getLongitude())
                .setLastUvIndex(locationInfo.getWeather().getUvi());
        } else {
            userEntity = new UserEntity()
                .setChatId(chatId)
                .setName(username)
                .setIsSubscribed(true)
                .setCreatedAt(Instant.now());
            LocationEntity newLocation = new LocationEntity()
                .setName(locationInfo.getName())
                .setLatitude(coordinates.getLatitude())
                .setLongitude(coordinates.getLongitude())
                .setLastUvIndex(locationInfo.getWeather().getUvi())
                .setUserEntity(userEntity)
                .setCreatedAt(Instant.now());
            userRepository.save(userEntity);
            locationRepository.save(newLocation);
        }
        return userEntity;
    }

    @Transactional
    public void setSubscription(Long chatId, boolean isSubscribed) {
        userRepository.getByChatId(chatId).setIsSubscribed(isSubscribed);
    }

    @Transactional
    public boolean isSubscribed(Long chatId) {
        return userRepository.findByChatId(chatId).map(UserEntity::getIsSubscribed).orElse(false);
    }

    @Transactional
    public void setLanguage(Long chatId, UserLanguage language) {
        userRepository.getByChatId(chatId).setLanguage(language);
    }
}

File: src/main/java/com/kmatrokhin/uvbot/services/UserService.kt
package com.kmatrokhin.uvbot.services

import com.kmatrokhin.uvbot.dto.Coordinates
import com.kmatrokhin.uvbot.dto.LocationInfo
import com.kmatrokhin.uvbot.dto.UserSignUp
import com.kmatrokhin.uvbot.entities.LocationEntity
import com.kmatrokhin.uvbot.entities.UserEntity
import com.kmatrokhin.uvbot.entities.UserLanguage
import com.kmatrokhin.uvbot.events.UserRegisteredEvent
import com.kmatrokhin.uvbot.repositories.LocationRepository
import com.kmatrokhin.uvbot.repositories.UserRepository
import org.springframework.context.ApplicationEventPublisher
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.Instant

@Service
class UserService(
    private val userRepository: UserRepository,
    private val locationRepository: LocationRepository,
    private val applicationEventPublisher: ApplicationEventPublisher
) {
    @Transactional
    fun signUpOrUpdate(userSignUp: UserSignUp): UserEntity {
        val chatId = userSignUp.chatId
        val username = userSignUp.name
        val locationInfo = userSignUp.locationInfo

        val coordinates = locationInfo.coordinates

        val existingUser = userRepository.findByChatId(chatId) ?: run {
            val userEntity = createUser(chatId, username)
            val newLocation = createLocation(userEntity, locationInfo)
            userRepository.save(userEntity)
            locationRepository.save(newLocation)
            applicationEventPublisher.publishEvent(UserRegisteredEvent(userEntity, newLocation))
            return userEntity
        }
        return updateExistingUser(existingUser, locationInfo, username, coordinates)
    }

    private fun createUser(chatId: Long, username: String?): UserEntity {
        return UserEntity(
            chatId = chatId,
            name = username,
            isSubscribed = true,
            createdAt = Instant.now()
        )
    }

    private fun createLocation(userEntity: UserEntity, locationInfo: LocationInfo): LocationEntity {
        return LocationEntity(
            name = locationInfo.name,
            latitude = locationInfo.coordinates.latitude,
            longitude = locationInfo.coordinates.longitude,
            lastUvIndex = locationInfo.weather.uvi,
            userEntity = userEntity,
            createdAt = Instant.now()
        )
    }

    private fun updateExistingUser(
        existingUser: UserEntity,
        locationInfo: LocationInfo,
        username: String?,
        coordinates: Coordinates
    ): UserEntity {
        existingUser.name = username
        existingUser.isSubscribed = true
        val locationEntity = locationRepository.findByUserEntity(existingUser)?.let {
            it.apply {
                name = locationInfo.name
                latitude = coordinates.latitude
                longitude = coordinates.longitude
                lastUvIndex = locationInfo.weather.uvi
            }
        }
        return existingUser
    }

    @Transactional
    fun setSubscription(chatId: Long, isSubscribed: Boolean) {
        userRepository.findByChatId(chatId)?.isSubscribed = isSubscribed
    }

    @Transactional
    fun isSubscribed(chatId: Long): Boolean {
        return userRepository.findByChatId(chatId)?.isSubscribed ?: false
    }

    @Transactional
    fun setLanguage(chatId: Long, language: UserLanguage) {
        userRepository.findByChatId(chatId)?.language = language
    }
}
